olp-cpp-sdk-core/src/network2/winhttp/NetworkProtocolWinHttp.cpp

```diff
@@ -0,0 +1,1063 @@
+    return olp::network2::ErrorCode::SUCCESS;
+  else if ((err == ERROR_WINHTTP_INVALID_URL) ||
+           (err == ERROR_WINHTTP_UNRECOGNIZED_SCHEME) ||
+           (err == ERROR_WINHTTP_NAME_NOT_RESOLVED))
+    return olp::network2::ErrorCode::INVALID_URL_ERROR;
+  else if ((err == ERROR_WINHTTP_CLIENT_AUTH_CERT_NEEDED) ||
+           (err == ERROR_WINHTTP_LOGIN_FAILURE) ||
+           (err == ERROR_WINHTTP_SECURE_FAILURE))
+    return olp::network2::ErrorCode::AUTHORIZATION_ERROR;
+  else if (err == ERROR_WINHTTP_OPERATION_CANCELLED)
+    return olp::network2::ErrorCode::CANCELLED_ERROR;
+  else if (err == ERROR_WINHTTP_TIMEOUT)
+    return olp::network2::ErrorCode::TIMEOUT_ERROR;
+  return olp::network2::ErrorCode::IO_ERROR;
+}
+
+LPWSTR
+queryHeadervalue(HINTERNET request, DWORD header) {
+  DWORD len = 0, index = WINHTTP_NO_HEADER_INDEX;
+  if (WinHttpQueryHeaders(request, header, WINHTTP_HEADER_NAME_BY_INDEX,
+                          WINHTTP_NO_OUTPUT_BUFFER, &len, &index))
+    return NULL;
+  if (GetLastError() != ERROR_INSUFFICIENT_BUFFER) {
+    return NULL;
+  }
+  LPWSTR buffer = (LPWSTR)LocalAlloc(LPTR, len);
+  if (!buffer) return NULL;
+  index = WINHTTP_NO_HEADER_INDEX;
+  if (!WinHttpQueryHeaders(request, header, WINHTTP_HEADER_NAME_BY_INDEX,
+                           buffer, &len, &index)) {
+    LocalFree(buffer);
+    return NULL;
+  }
+  return buffer;
+}
+
+void UnixTimeToFileTime(std::uint64_t t, LPFILETIME pft) {
+  // Note that LONGLONG is a 64-bit value
+  LONGLONG ll;
+
+  ll = Int32x32To64(t, 10000000) + 116444736000000000;
+  pft->dwLowDateTime = (DWORD)ll;
+  pft->dwHighDateTime = ll >> 32;
+}
+
+bool convertMultiByteToWideChar(const std::string& s_in, std::wstring& s_out) {
+  s_out.clear();
+  if (s_in.empty()) {
+    return true;
+  }
+
+  // Detect required buffer size.
+  const auto chars_required = MultiByteToWideChar(
+      CP_ACP, MB_PRECOMPOSED, s_in.c_str(),
+      -1,       // denotes null-terminated string
+      nullptr,  // output buffer is null means request required buffer size
+      0);
+
+  if (chars_required == 0) {
+    // error: could not convert input string
+    return false;
+  }
+
+  s_out.resize(chars_required);
+  // Perform actual conversion from multi-byte to wide char
+  const auto conversion_result =
+      MultiByteToWideChar(CP_ACP, MB_PRECOMPOSED, s_in.c_str(),
+                          -1,  // denotes null-terminated string
+                          &s_out.front(), s_out.size());
+
+  if (conversion_result == 0) {
+    // Should not happen as 1st call have succeeded.
+    return false;
+  }
+  return true;
+}
+}  // namespace
+
+namespace olp {
+namespace network2 {
+
+#define LOGTAG "WinHttp"
+
+NetworkProtocolWinHttp::NetworkProtocolWinHttp()
+    : http_session_(NULL),
+      thread_(INVALID_HANDLE_VALUE),
+      event_(INVALID_HANDLE_VALUE) {
+  recent_request_id_.store(1);
+  http_session_ = WinHttpOpen(L"NGMOS CLient", WINHTTP_ACCESS_TYPE_NO_PROXY,
```

Comment:
```suggestion
  http_session_ = WinHttpOpen(L"OLP Http Client", WINHTTP_ACCESS_TYPE_NO_PROXY,
```
Is this valid name? I guess current name uses name of the legacy HERE project...